ext/prism/extconf.rb

```diff
@@ -1,136 +1,101 @@
-      end
-
-      def print_help
-        print(<<~TEXT)
-          USAGE: ruby #{$PROGRAM_NAME} [options]
-
-            Flags that are always valid:
-
-                --enable-static
-                --disable-static
-                    Enable or disable static linking against librubyparser.
-                    The default is to statically link.
-
-                --enable-debug-mode-build
-                    Enable debug mode build.
-                    You may also use set PRISM_DEBUG_MODE_BUILD environment variable.
-
-                --help
-                    Display this message.
-
-            Environment variables used:
-
-                PRISM_DEBUG_MODE_BUILD
-                    Equivalent to `--enable-debug-mode-build` when set, even if nil or blank.
-
-        TEXT
-      end
-
-      def static_link?
-        enable_config("static", true)
-      end
-
-      def debug_mode_build?
-        enable_config("debug-mode-build", ENV["PRISM_DEBUG_MODE_BUILD"] || false)
-      end
+          PRISM_DEBUG_MODE_BUILD
+              Equivalent to `--enable-debug-mode-build` when set, even if nil or blank.
+
+  TEXT
+  exit!(0)
+end
+
+# Runs `make` in the root directory of the project. Note that this is the
+# `Makefile` for the overall project, not the `Makefile` that is being generated
+# by this script.`
+def make(target)
+  Dir.chdir(File.expand_path("../..", __dir__)) do
+    # If this gem is being build from a git source, then we need to run
+    # templating if it hasn't been run yet. In normal packaging, we would have
+    # shipped the templated files with the gem, so this wouldn't be necessary.
+    if !File.exist?("include/prism/ast.h") && Dir.exist?(".git")
+      system("templates/template.rb", exception: true)
     end
+
+    system("make", target, exception: true)
   end
 end
 
-if ARGV.delete("--help")
-  Prism::ExtConf.print_help
-  exit!(0)
+require "rbconfig"
+
+# On non-CRuby we only need the shared library since we'll interface with it
+# through FFI, so we'll build only that and not the C extension. We also avoid
+# `require "mkmf"` as that prepends the LLVM toolchain to PATH on TruffleRuby,
+# but we want to use the native toolchain here since librubyparser is run
+# natively.
+if RUBY_ENGINE != "ruby"
+  make("build/librubyparser.#{RbConfig::CONFIG["SOEXT"]}")
+  File.write("Makefile", "all install clean:\n\t@#{RbConfig::CONFIG["NULLCMD"]}\n")
+  return
+end
+
+require "mkmf"
+
+# First, ensure that we can find the header for the prism library.
+unless find_header("prism.h", File.expand_path("../../include", __dir__))
+  raise "prism.h is required"
 end
 
-Prism::ExtConf.configure
+# Next, ensure we can find the header for the C extension. Explicitly look for
+# the extension header in the parent directory because we want to consistently
+# look for `prism/extension.h` in our source files to line up with our mirroring
+# in CRuby.
+unless find_header("prism/extension.h", File.join(__dir__, ".."))
```

Comment:
FWIW File.join just adds `/` even on Windows, so it's the same as `"#{__dir__}/.."` (which IMO is clearer, but subjective).
Alternatively `File.expand_path("..", __dir__)` like above might be nicer as it resolves the `..`.